File: frontend/src/types/job.ts
export interface JobPosting {
  id: string
  company_id: string
  title: string
  description: string
  summary?: string
  job_type: JobType
  experience_level: string
  department?: string
  location?: string
  remote_type: RemoteType
  salary_min?: number
  salary_max?: number
  salary_currency: string
  benefits: string[]
  requirements?: string
  responsibilities?: string
  qualifications?: string
  application_deadline?: string
  max_applications?: number
  application_instructions?: string
  status: JobStatus
  is_featured: boolean
  is_urgent: boolean
  slug?: string
  meta_description?: string
  tags: string[]
  posted_at?: string
  expires_at?: string
  created_at: string
  updated_at: string
  view_count: number
  application_count: number
  company?: Company
  required_skills?: Skill[]
  applications?: JobApplication[]
}

export interface JobApplication {
  id: string
  job_posting_id: string
  candidate_id: string
  status: ApplicationStatus
  cover_letter?: string
  resume_url?: string
  screening_questions?: string
  ai_match_score?: number
  recruiter_notes?: string
  applied_at: string
  reviewed_at?: string
  status_updated_at: string
  last_contact_at?: string
  next_followup_at?: string
  created_at: string
  updated_at: string
  job_posting?: JobPosting
  candidate?: CandidateProfile
}

export interface CandidateProfile {
  id: string
  user_id: string
  first_name: string
  last_name: string
  email: string
  resume_url?: string
  skills: Skill[]
  experience_years: number
  education: Education[]
  certifications: Certification[]
  preferred_locations: string[]
  salary_expectation?: SalaryRange
  availability?: string
  profile_picture?: string
  bio?: string
}

export interface Company {
  id: string
  company_name: string
  industry: string
  company_size: string
  website?: string
  description?: string
  logo_url?: string
  verified: boolean
}

export interface Skill {
  id: string
  name: string
  category: string
  level?: string
}

export interface Education {
  id: string
  institution: string
  degree: string
  field_of_study: string
  start_date: string
  end_date?: string
  grade?: string
}

export interface Certification {
  id: string
  name: string
  issuing_organization: string
  issue_date: string
  expiration_date?: string
  credential_id?: string
  credential_url?: string
}

export interface SalaryRange {
  min: number
  max: number
  currency: string
}

export enum JobType {
  FULL_TIME = "full_time",
  PART_TIME = "part_time",
  CONTRACT = "contract",
  TEMPORARY = "temporary",
  INTERNSHIP = "internship",
  FREELANCE = "freelance"
}

export enum RemoteType {
  ONSITE = "onsite",
  REMOTE = "remote",
  HYBRID = "hybrid"
}

export enum JobStatus {
  DRAFT = "draft",
  ACTIVE = "active",
  PAUSED = "paused",
  EXPIRED = "expired",
  FILLED = "filled",
  CANCELLED = "cancelled"
}

export enum ApplicationStatus {
  PENDING = "pending",
  REVIEWING = "reviewing",
  SHORTLISTED = "shortlisted",
  INTERVIEWED = "interviewed",
  OFFERED = "offered",
  ACCEPTED = "accepted",
  REJECTED = "rejected",
  WITHDRAWN = "withdrawn"
}

export interface JobSearchFilters {
  query?: string
  location?: string
  remote_type?: RemoteType[]
  job_type?: JobType[]
  experience_level?: string[]
  salary_min?: number
  salary_max?: number
  skills?: string[]
  company_size?: string[]
  posted_within?: number // days
  sort_by?: 'relevance' | 'date' | 'salary' | 'company'
  sort_order?: 'asc' | 'desc'
}

export interface JobSearchResponse {
  jobs: JobPosting[]
  total: number
  page: number
  per_page: number
  total_pages: number
  filters_applied: JobSearchFilters
}

File: backend/app/models/job.py
from sqlalchemy import Column, String, Integer, Float, Boolean, DateTime, Text, ForeignKey, Table
from sqlalchemy.dialects.postgresql import UUID, ARRAY
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from enum import Enum
import uuid
from ..database import Base


class JobStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    EXPIRED = "expired"
    FILLED = "filled"
    CANCELLED = "cancelled"


class JobType(str, Enum):
    FULL_TIME = "full_time"
    PART_TIME = "part_time"
    CONTRACT = "contract"
    TEMPORARY = "temporary"
    INTERNSHIP = "internship"
    FREELANCE = "freelance"


class RemoteType(str, Enum):
    ONSITE = "onsite"
    REMOTE = "remote"
    HYBRID = "hybrid"


class ApplicationStatus(str, Enum):
    PENDING = "pending"
    REVIEWING = "reviewing"
    SHORTLISTED = "shortlisted"
    INTERVIEWED = "interviewed"
    OFFERED = "offered"
    ACCEPTED = "accepted"
    REJECTED = "rejected"
    WITHDRAWN = "withdrawn"


# Association table for job required skills
job_skills = Table(
    'job_skills',
    Base.metadata,
    Column('job_posting_id', UUID(as_uuid=True), ForeignKey('job_postings.id')),
    Column('skill_id', UUID(as_uuid=True), ForeignKey('skills.id'))
)


class JobPosting(Base):
    __tablename__ = "job_postings"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    company_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
    
    # Basic Information
    title = Column(String, nullable=False, index=True)
    description = Column(Text, nullable=False)
    summary = Column(Text, nullable=True)  # Short summary for listings
    
    # Job Details
    job_type = Column(String, default=JobType.FULL_TIME)
    experience_level = Column(String, nullable=False)  # From profile.py ExperienceLevel
    department = Column(String, nullable=True)
    
    # Location and Remote
    location = Column(String, nullable=True)
    remote_type = Column(String, default=RemoteType.HYBRID)
    
    # Compensation
    salary_min = Column(Integer, nullable=True)
    salary_max = Column(Integer, nullable=True)
    salary_currency = Column(String, default="USD")
    benefits = Column(ARRAY(String), default=[])
    
    # Requirements
    requirements = Column(Text, nullable=True)
    responsibilities = Column(Text, nullable=True)
    qualifications = Column(Text, nullable=True)
    
    # Application Settings
    application_deadline = Column(DateTime(timezone=True), nullable=True)
    max_applications = Column(Integer, nullable=True)
    application_instructions = Column(Text, nullable=True)
    
    # Status and Visibility
    status = Column(String, default=JobStatus.DRAFT, index=True)
    is_featured = Column(Boolean, default=False)
    is_urgent = Column(Boolean, default=False)
    
    # SEO and Metadata
    slug = Column(String, unique=True, nullable=True, index=True)
    meta_description = Column(String, nullable=True)
    tags = Column(ARRAY(String), default=[])
    
    # Timestamps
    posted_at = Column(DateTime(timezone=True), nullable=True)
    expires_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Analytics
    view_count = Column(Integer, default=0)
    application_count = Column(Integer, default=0)
    
    # Relationships
    company = relationship("User", back_populates="job_postings")
    required_skills = relationship("Skill", secondary=job_skills, back_populates="job_postings")
    applications = relationship("JobApplication", back_populates="job_posting", cascade="all, delete-orphan")


class JobApplication(Base):
    __tablename__ = "job_applications"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    job_posting_id = Column(UUID(as_uuid=True), ForeignKey('job_postings.id'), nullable=False)
    candidate_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
    
    # Application Details
    status = Column(String, default=ApplicationStatus.PENDING, index=True)
    cover_letter = Column(Text, nullable=True)
    resume_url = Column(String, nullable=True)  # Can override candidate's default resume
    
    # Screening
    screening_questions = Column(Text, nullable=True)  # JSON string of questions and answers
    ai_match_score = Column(Float, nullable=True)  # AI-calculated compatibility score
    recruiter_notes = Column(Text, nullable=True)
    
    # Timeline
    applied_at = Column(DateTime(timezone=True), server_default=func.now())
    reviewed_at = Column(DateTime(timezone=True), nullable=True)
    status_updated_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Communication
    last_contact_at = Column(DateTime(timezone=True), nullable=True)
    next_followup_at = Column(DateTime(timezone=True), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    job_posting = relationship("JobPosting", back_populates="applications")
    candidate = relationship("User", foreign_keys=[candidate_id], back_populates="applications")


class SavedJob(Base):
    __tablename__ = "saved_jobs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    candidate_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
    job_posting_id = Column(UUID(as_uuid=True), ForeignKey('job_postings.id'), nullable=False)
    
    # Metadata
    notes = Column(Text, nullable=True)
    saved_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    candidate = relationship("User", back_populates="saved_jobs")
    job_posting = relationship("JobPosting")
    
    # Ensure unique constraint
    __table_args__ = (
        {'extend_existing': True},
    )


# Add back_populates to existing models
from .profile import Skill
Skill.job_postings = relationship("JobPosting", secondary=job_skills, back_populates="required_skills")